src/sentry/integrations/vercel/uninstall.py

```diff
@@ -0,0 +1,96 @@
+
+from django.views.decorators.csrf import csrf_exempt
+from sentry.api.base import Endpoint
+from sentry.models import AuditLogEntryEvent, Integration, OrganizationIntegration, Organization
+from sentry.utils.audit import create_audit_entry
+from sentry.web.decorators import transaction_start
+
+logger = logging.getLogger("sentry.integrations.vercel.uninstall")
+
+
+class VercelUninstallEndpoint(Endpoint):
+    authentication_classes = ()
+    permission_classes = ()
+
+    @csrf_exempt
+    def dispatch(self, request, *args, **kwargs):
+        return super(VercelUninstallEndpoint, self).dispatch(request, *args, **kwargs)
+
+    @transaction_start("VercelUninstallEndpoint")
+    def post(self, request):
+        # userId should always be present
+        external_id = request.data.get("teamId") or request.data.get("userId")
+        configuration_id = request.data.get("configurationId")
+
+        try:
+            integration = Integration.objects.get(provider="vercel", external_id=external_id)
+        except Integration.DoesNotExist:
+            logger.info(
+                "vercel.uninstall.missing-integration",
+                extra={"configuration_id": configuration_id, "external_id": external_id},
+            )
+            return self.respond(status=404)
+
+        orgs = integration.organizations.all()
+
+        if len(orgs) == 1:
+            create_audit_entry(
+                request=request,
+                organization=integration.organizations.all()[0],
+                target_object=integration.id,
+                event=AuditLogEntryEvent.INTEGRATION_REMOVE,
+                # TODO(meredith): If we create vercel identities from the userId
+                # we could attempt to find the user in Sentry and pass that in for
+                # the actor instead.
+                actor_label="Vercel User",
+                data={"provider": integration.provider, "name": integration.name},
+            )
+            integration.delete()
+            return self.respond(status=202)
+
+        configuration = integration.metadata["configurations"].pop(configuration_id)
+        if configuration_id == integration.metadata["installation_id"]:
+            # if we are uninstalling a primary configuration, and there are
+            # multiple orgs connected to this integration we must update
+            # the crendentials (access_token, webhook_id etc)
+            next_config_id, next_config = integration.metadata["configurations"].items()[0]
+
+            integration.metadata["access_token"] = next_config["access_token"]
+            integration.metadata["webhook_id"] = next_config["webhook_id"]
+            integration.metadata["installation_id"] = next_config_id
+
+        try:
+            OrganizationIntegration.objects.get(
+                organization_id=configuration["organization_id"], integration_id=integration.id
+            ).delete()
+        except OrganizationIntegration.DoesNotExist:
+            logger.info(
```

Comment:
@MeredithAnya This should probably be an error, yea? I can't imagine when this would happen except if there is a bug.

Reply:
okays yes, since we are going to be forcing people to uninstall from Vercel.